src/main/java/org/elasticsearch/plugins/repository/LocalPluginRepository.java

```diff
@@ -0,0 +1,433 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugins.repository;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.FileSystemUtils;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.plugins.PluginsService;
+
+import java.io.IOException;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.PosixFileAttributeView;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.Collection;
+import java.util.Iterator;
+
+import static org.elasticsearch.plugins.repository.PluginDescriptor.Builder.pluginDescriptor;
+
+/**
+ * Local plugin repository used to manage installed plugins.
+ */
+public class LocalPluginRepository implements PluginRepository {
+
+    public static final ImmutableSet<String> FORBIDDEN_NAMES = ImmutableSet.of("elasticsearch", "elasticsearch.bat", "elasticsearch.in.sh",
+                                                                                "plugin", "plugin.bat", "service.bat");
+
+    private final static String BIN = "bin";
+    private final static String CONFIG = "config";
+    private final static String SITE = "_site";
+    private final static String JAR = ".jar";
+    private final static String CLASS = ".class";
+    private final static String JAVA = ".java";
+
+    private final String name;
+
+    private final Path pluginsDir;
+    private final Path binDir;
+    private final Path configDir;
+
+    private final Settings settings;
+
+    private LocalPluginRepository(String name, Environment env, Settings settings) {
+        this.name = name;
+
+        this.binDir = env.homeFile().resolve(BIN);
```

Comment:
Not really related to your PR but I think we should may be add a `binFile()` method in `Environment`. Do you think this is something doable within your change?